kube-controllers/pkg/controllers/node/hostendpoints.go

```diff
@@ -17,300 +17,425 @@ package node
-}
-
-// syncAutoHostendpointWithRetries syncs the auto hostendpoint for the given
-// node, retrying a few times if needed.
-func (c *autoHostEndpointController) syncAutoHostendpointWithRetries(ctx context.Context, node *libapi.Node) error {
-	for n := 1; n <= 5; n++ {
-		logrus.Debugf("syncing hostendpoint for node %q. attempt #%v", node.Name, n)
-		if err := c.syncAutoHostendpoint(ctx, node); err != nil {
-			logrus.WithError(err).Infof("failed to sync host endpoint for node %q, retrying", node.Name)
-			time.Sleep(retrySleepTime)
-			continue
+	// Check that there are no lingering hostEndpoints that no longer match the template spec for this node
+	// We want to delete all HostEndpoints in our HostEndpointTracker that are not part of hostEndpointsMatchingNode
+	for _, hostEndpoint := range c.hostEndpointTracker.getHostEndpointsForNode(node.Name) {
+		if _, ok := hostEndpointsMatchingNode[hostEndpoint.Name]; !ok {
+			logrus.Infof("hostEndpoint %s no longer matches template", hostEndpoint.Name)
+			err := c.deleteHostEndpoint(hostEndpoint.Name)
+			if err != nil {
+				logrus.WithError(err).Error("failed to delete host endpoint")
+			}
 		}
-		return nil
 	}
-	return fmt.Errorf("too many retries when syncing hostendpoint for node %q", node.Name)
 }
 
-// listAutoHostendpoints returns a map of auto hostendpoints keyed by the
-// hostendpoint's name.
-func (c *autoHostEndpointController) listAutoHostendpoints(ctx context.Context) (map[string]api.HostEndpoint, error) {
-	time.Sleep(c.rl.When(RateLimitCalicoList))
-	heps, err := c.client.HostEndpoints().List(ctx, options.ListOptions{})
-	if err != nil {
-		return nil, fmt.Errorf("could not list hostendpoints: %v", err.Error())
-	}
-	c.rl.Forget(RateLimitCalicoList)
-	m := make(map[string]api.HostEndpoint)
-	for _, h := range heps.Items {
-		if isAutoHostendpoint(&h) {
-			m[h.Name] = h
+// deleteHostEndpointsForNode removes all HostEndpoints associated with the Node
+func (c *autoHostEndpointController) deleteHostEndpointsForNode(nodeName string) {
+	for _, hep := range c.hostEndpointTracker.getHostEndpointsForNode(nodeName) {
+		err := c.deleteHostEndpoint(hep.Name)
+		if err != nil {
+			logrus.WithError(err).Error("failed to delete host endpoint")
 		}
 	}
-	return m, nil
 }
 
-// deleteHostendpoint removes the specified hostendpoint, optionally retrying
-// the operation a few times until it succeeds.
-func (c *autoHostEndpointController) deleteHostendpoint(ctx context.Context, hepName string) error {
+// deleteHostEndpoint removes the specified HostEndpoint
+func (c *autoHostEndpointController) deleteHostEndpoint(hepName string) error {
 	logrus.Debugf("deleting hostendpoint %q", hepName)
-	rlKey := rateLimiterItemKey{Type: RateLimitCalicoDelete, Name: hepName}
-	time.Sleep(c.rl.When(rlKey))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
 	_, err := c.client.HostEndpoints().Delete(ctx, hepName, options.DeleteOptions{})
 	if err != nil {
```

Comment:
It's often a good idea to have a `IsNotExists` check here - if we attempt to delete something and it doesn't exist (e.g., because someone else deleted it before we coudl), we don't need to treat that as an error